### DISCUSSION TITLE
trouble installing - docker compse / portainer

### DISCUSSION THREAD

**Discussion by JamesM92** [mastodon/mastodon#18250] [Q&A]
I've been trying for days to get a mastodon server up and running on my home lab but i keep running into issues i can't seem to resolve.

I'm trying to do this with docker compose, and mostly through my portainer manager if i can.

My server already has a established postgresql and redis servers, that id want to have it try to use if i could . (i recently made the shared databases as i had multiple services using their own and it was causing conflicts)

I haven't seen any clear instructions on what to do in this scenario.

1. installing on server with already established postgres and redis servers
2. open a new portainer stack editor 
3. copy $ paste docker-compose.yml file into portainer editor, commenting out details for the postgres and redis services
4. copy .env.production.sample as .env.production in the location the compose file is located
5.  via command line navigate to where the docker-compose file is lcoated and run 
"docker-compose run --rm web bundle exec rake mastodon:setup"
6. walk-through the setup instructions.
7. once complete make sure to stop all the mastodon services, then re-deploy the stack
8. 
...

### Expected behaviour

services to function

### Actual behaviour

services have errors

### Specifications

mastodon:v3.5.1
postgres:14-alpine
redis:6-alpine
docker: 5:20.10.13
portainer.io:2.11.0

error logs:
**_### Mastodon_web:_**
`=> Booting Puma
=> Rails 6.1.5 application starting in production 
=> Run `bin/rails server --help` for more startup options
Exiting
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/application.rb:601:in `validate_secret_key_base': `secret_key_base` for production environment must be a type of String` (ArgumentError)
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/application.rb:437:in `secret_key_base'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/devise-4.8.1/lib/devise/secret_key_finder.rb:24:in `key_exists?'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/devise-4.8.1/lib/devise/secret_key_finder.rb:16:in `find'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/devise-4.8.1/lib/devise/rails.rb:37:in `block in <class:Engine>'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:32:in `instance_exec'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:32:in `run'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:61:in `block in run_initializers'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:228:in `block in tsort_each'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:431:in `each_strongly_connected_component_from'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:349:in `block in each_strongly_connected_component'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:347:in `each'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:347:in `call'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:347:in `each_strongly_connected_component'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:226:in `tsort_each'
	from /opt/ruby/lib/ruby/3.0.0/tsort.rb:205:in `tsort_each'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:60:in `run_initializers'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/application.rb:391:in `initialize!'
	from /opt/mastodon/config/environment.rb:5:in `<top (required)>'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/bootsnap-1.10.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/bootsnap-1.10.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:35:in `require'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.5/lib/active_support/dependencies.rb:332:in `block in require'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.5/lib/active_support/dependencies.rb:299:in `load_dependency'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.5/lib/active_support/dependencies.rb:332:in `require'
	from config.ru:4:in `block in <main>'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/builder.rb:116:in `eval'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/builder.rb:116:in `new_from_string'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/builder.rb:105:in `load_file'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/builder.rb:66:in `parse_file'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/server.rb:349:in `build_app_and_options_from_config'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/server.rb:249:in `app'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/server.rb:422:in `wrapped_app'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/server.rb:312:in `block in start'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/server.rb:379:in `handle_profiling'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/server.rb:311:in `start'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/commands/server/server_command.rb:39:in `start'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/commands/server/server_command.rb:144:in `block in perform'
	from <internal:kernel>:90:in `tap'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/commands/server/server_command.rb:135:in `perform'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/command/base.rb:69:in `perform'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/command.rb:48:in `invoke'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/commands.rb:18:in `<top (required)>'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/bootsnap-1.10.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/bootsnap-1.10.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from bin/rails:4:in `<main>'`

_**### mastodon_stream:**_
ERR! Redis Client Error! Error: connect ECONNREFUSED 127.0.0.1:6379
ERR! Redis Client Error!     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1157:16)
ERR! Redis Client Error!  Error: connect ECONNREFUSED 127.0.0.1:6379
ERR! Redis Client Error!     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1157:16) {
ERR! Redis Client Error!   errno: -111,
ERR! Redis Client Error!   code: 'ECONNREFUSED',
ERR! Redis Client Error!   syscall: 'connect',
ERR! Redis Client Error!   address: '127.0.0.1',
ERR! Redis Client Error!   port: 6379
ERR! Redis Client Error! }

_**### mastodon_sidekik:**_
2022-05-02T12:52:38.461Z pid=7 tid=6zn INFO: Booting Sidekiq 6.4.1 with redis options {:driver=>:hiredis, :url=>"redis://localhost:6379/0", :namespace=>nil}
2022-05-02T12:52:38.535Z pid=7 tid=6zn WARN: ArgumentError: `secret_key_base` for production environment must be a type of String`
2022-05-02T12:52:38.535Z pid=7 tid=6zn WARN: /opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/application.rb:601:in `validate_secret_key_base'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/application.rb:437:in `secret_key_base'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/devise-4.8.1/lib/devise/secret_key_finder.rb:24:in `key_exists?'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/devise-4.8.1/lib/devise/secret_key_finder.rb:16:in `find'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/devise-4.8.1/lib/devise/rails.rb:37:in `block in <class:Engine>'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:32:in `instance_exec'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:32:in `run'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:61:in `block in run_initializers'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:228:in `block in tsort_each'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:431:in `each_strongly_connected_component_from'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:349:in `block in each_strongly_connected_component'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:347:in `each'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:347:in `call'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:347:in `each_strongly_connected_component'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:226:in `tsort_each'
/opt/ruby/lib/ruby/3.0.0/tsort.rb:205:in `tsort_each'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/initializable.rb:60:in `run_initializers'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/railties-6.1.5/lib/rails/application.rb:391:in `initialize!'
/opt/mastodon/config/environment.rb:5:in `<top (required)>'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.5/lib/active_support/dependencies.rb:332:in `require'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.5/lib/active_support/dependencies.rb:332:in `block in require'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.5/lib/active_support/dependencies.rb:299:in `load_dependency'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.5/lib/active_support/dependencies.rb:332:in `require'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/sidekiq-6.4.1/lib/sidekiq/cli.rb:273:in `boot_application'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/sidekiq-6.4.1/lib/sidekiq/cli.rb:37:in `run'
/opt/mastodon/vendor/bundle/ruby/3.0.0/gems/sidekiq-6.4.1/bin/sidekiq:31:in `<top (required)>'
/opt/mastodon/vendor/bundle/ruby/3.0.0/bin/sidekiq:25:in `load'
/opt/mastodon/vendor/bundle/ruby/3.0.0/bin/sidekiq:25:in `<top (required)>'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/cli/exec.rb:58:in `load'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/cli/exec.rb:58:in `kernel_load'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/cli/exec.rb:23:in `run'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/cli.rb:483:in `exec'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/cli.rb:31:in `dispatch'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/cli.rb:25:in `start'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/exe/bundle:48:in `block in <top (required)>'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'
/opt/ruby/lib/ruby/gems/3.0.0/gems/bundler-2.3.11/exe/bundle:36:in `<top (required)>'
/opt/ruby/bin/bundle:23:in `load'
/opt/ruby/bin/bundle:23:in `<main>'

**Comment by david-vega** [comment]
Copying the `.env.production.sample` as `.env.production` is not enough. You also need to modify the file. it seems that you are missing to set values like `secret_key_base`. As it says in the `.env.production.sample` you have to: "Make sure to use `rake secret` to generate secrets". 

Once you set all the values required it should work :)
